src/object.jl

```diff
@@ -196,6 +196,13 @@ mutable struct PrefetchedDownloadStream{T <: Object} <: IO
     )
         url = makeURL(store, key)
         resp = API.headObject(store, url, HTTP.Headers(); credentials=credentials)
+        if HTTP.isredirect(resp)
+            try
+                throw(HTTP.StatusError(resp.status, resp.request.method, resp.request.target, resp))
+            catch
+                throw(ArgumentError("Invalid object key: $key"))
+            end
```

Comment:
Can you explain this with a code comment? How could throwing the StatusError exception cause another exception to be raised?
Actually, wait, won't the throw from the try _always_ end up in the catch? I don't understand this

Reply:
And what does any of that have to do with a redirect? 😅

Reply:
Per Jacob's commit message introducing this pattern:
> The issue is that HTTP.jl doesn't consider redirect responses as
errors by default (they're usually handled automatically via default
redirect logic, except for 301 statuses, which means the object has
moved permanently and no redirect location is known).
This PR checks for this case and throw an appropriate ArgumentError.

I guess the second throw could happen if there is no `resp.request` object. Perhaps we should wait for Jacob to confirm and potentially clean this pattern a bit in all the callsites in this package.

Reply:
This is a pattern you can use to "chain" exceptions together. i.e. we want to throw a CloudBase-specific error (ArgumentError, invalid object), but we also want to provide the context of the HTTP status code, that it was a redirect. So we first throw a HTTP.StatusError (mainly, as mentioned, HTTP doesn't throw this by default), and then throw the CloudBase error, which has the effect of showing both errors like:
```
julia> S3.get("s3://real-bucket/fake-object.csv")
┌ Warning: `region` keyword argument not provided to `S3.get` and undetected from url.  Defaulting to `us-east-1`
└ @ CloudStore.S3 ~/.julia/dev/CloudStore/src/s3.jl:68
ERROR: ArgumentError: Invalid object key: fake-object.csv
Stacktrace:
 [1] getObjectImpl(x::CloudBase.AWS.Bucket, key::String, out::Nothing; multipartThreshold::Int64, partSize::Int64, batchSize::Int64, allowMultipart::Bool, decompress::Bool, headers::Vector{Pair{SubString{String}, SubString{String}}}, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ CloudStore.API ~/.julia/dev/CloudStore/src/get.jl:62
 [2] getObjectImpl (repeats 2 times)
   @ ~/.julia/dev/CloudStore/src/get.jl:45 [inlined]
 [3] #get#7
   @ ~/.julia/dev/CloudStore/src/s3.jl:33 [inlined]
 [4] get
   @ ~/.julia/dev/CloudStore/src/s3.jl:33 [inlined]
 [5] get(::String; region::Nothing, nowarn::Bool, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ CloudStore.S3 ~/.julia/dev/CloudStore/src/s3.jl:72
 [6] get(::String)
   @ CloudStore.S3 ~/.julia/dev/CloudStore/src/s3.jl:64
 [7] top-level scope
   @ REPL[44]:1

caused by: HTTP.Exceptions.StatusError(301, "HEAD", "/fake-object.csv", HTTP.Messages.Response:
"""
HTTP/1.1 301 Moved Permanently
x-amz-bucket-region: us-west-1
x-amz-request-id: N0CNYH71Z07VX1NY
x-amz-id-2: YI0zmB3s+tnHjn0D6jNOTGlP5EF7aP1POjnnJWCnNVsd+wsc0TTWuDeAMo2CbfdVUDYtEE81oGU=
Content-Type: application/xml
Date: Wed, 18 Jan 2023 22:18:08 GMT
Server: AmazonS3

""")
Stacktrace:
 [1] getObjectImpl(x::CloudBase.AWS.Bucket, key::String, out::Nothing; multipartThreshold::Int64, partSize::Int64, batchSize::Int64, allowMultipart::Bool, decompress::Bool, headers::Vector{Pair{SubString{String}, SubString{String}}}, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ CloudStore.API ~/.julia/dev/CloudStore/src/get.jl:60
 [2] getObjectImpl (repeats 2 times)
   @ ~/.julia/dev/CloudStore/src/get.jl:45 [inlined]
 [3] #get#7
   @ ~/.julia/dev/CloudStore/src/s3.jl:33 [inlined]
 [4] get
   @ ~/.julia/dev/CloudStore/src/s3.jl:33 [inlined]
 [5] get(::String; region::Nothing, nowarn::Bool, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ CloudStore.S3 ~/.julia/dev/CloudStore/src/s3.jl:72
 [6] get(::String)
   @ CloudStore.S3 ~/.julia/dev/CloudStore/src/s3.jl:64
 [7] top-level scope
   @ REPL[44]:1
```

Reply:
Cool, thanks! I was wondering if it was that. Neat.

I definitely think this deserves a quick code-comment. That's an unusual use of exceptions, and i think that it would benefit from a short one-line comment explaining what it's doing. 😊

Reply:
I've added the comment, thanks both, TIL:) Gonna merge this now